sound/soc/sof/intel/hda-dai.c

```diff
@@ -168,20 +102,21 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
 	struct hdac_bus *bus;
 	unsigned int format_val;
 	unsigned int link_bps;
-	unsigned char stream_tag;
+	int stream_tag;
 
 	sdev = snd_soc_component_get_drvdata(cpu_dai->component);
 	bus = sof_to_bus(sdev);
 
-	hext_stream = snd_soc_dai_get_dma_data(cpu_dai, substream);
-	if (!hext_stream) {
-		hext_stream = hda_link_stream_assign(bus, substream);
-		if (!hext_stream)
-			return -EBUSY;
+	hext_stream = ops->get_hext_stream(sdev, cpu_dai, substream);
 
-		snd_soc_dai_set_dma_data(cpu_dai, substream, (void *)hext_stream);
+	if (!hext_stream) {
+		if (ops->assign_hext_stream)
+			hext_stream = ops->assign_hext_stream(sdev, cpu_dai, substream);
 	}
```

Comment:
You got me here...
Only `get_hext_stream` is mandatory, which is to to retrieve the dma_data set by the optional `assign_hext_stream`.
If the `assign_hext_stream` is not set (optional!) and this is the first call here then the `get_hext_stream` is going to return with NULL and you are not going to get anything but NULL...

What I was saying is that most, if not all callbacks in `struct hda_dai_widget_dma_ops` are in fact mandatory, don't they?

Reply:
also to note: you also must have the `release_hext_stream` defined if you have the `assign_hext_stream`, but you must need to have the `assign_hext_stream` in order to have working stack.

Reply:
or you should not return with error in line 118?

Reply:
The conclusion for this thread is that assign/release need to be optional because couple mode will not need them